pallets/name-service/src/lib.rs

```diff
@@ -0,0 +1,445 @@
+            frame_system::Pallet::<T>::block_number() > pending_register_data.1,
+            Error::<T>::RegisterTimeNotReached
+        );
+
+        ensure!(
+            pending_register_data.0 == hash_address,
+            Error::<T>::NotOwned
+        );
+
+        // Move from pending into records
+        PendingRegister::<T>::remove(hash_user);
+        UsernameRecords::<T>::insert(&username, registrant);
+
+        Self::deposit_event(Event::NameRegistered {
+            username,
+            owner: registrant,
+        });
+        Ok(())
+    }
+
+    /// Set primary name if register and owned
+    fn try_set_primary_name(username: UserName, registrant: ZkAddressType) -> DispatchResult {
+        // Username checks
+        username_validation(&username).ok_or(Error::<T>::InvalidUsernameFormat)?;
+
+        //check if name is registered
+        ensure!(
+            UsernameRecords::<T>::contains_key(&username),
+            Error::<T>::NotRegistered
+        );
+        // check if its owned (can unwrap because of previous checks for the key)
+        ensure!(
+            UsernameRecords::<T>::get(&username).unwrap() == registrant,
+            Error::<T>::NotOwned
+        );
+
+        // check if we already have a primary
+        if PrimaryRecords::<T>::contains_key(registrant) {
+            PrimaryRecords::<T>::mutate(registrant, |old_username| {
+                *old_username = Some(username.clone())
+            });
+        } else {
+            PrimaryRecords::<T>::insert(registrant, &username);
+        }
+
+        Self::deposit_event(Event::NameSetAsPrimary {
+            owner: registrant,
+            username,
+        });
+        Ok(())
+    }
+
+    fn try_cancel_pending_register(
+        username: UserName,
+        registrant: ZkAddressType,
+    ) -> DispatchResult {
+        // Username checks
+        username_validation(&username).ok_or(Error::<T>::InvalidUsernameFormat)?;
+
+        let (hash_user, hash_address) = (
+            T::Hashing::hash_of(&username),
+            T::Hashing::hash_of(&registrant),
+        );
+
+        let pending_register_data =
+            PendingRegister::<T>::get(hash_user).ok_or(Error::<T>::UsernameNotFound)?;
+
+        ensure!(
+            pending_register_data.0 == hash_address,
+            Error::<T>::NotOwned
+        );
+
+        PendingRegister::<T>::remove(hash_user);
+
+        Self::deposit_event(Event::RegisterCanceled {
+            hash_username: hash_user,
+            hash_owner: hash_address,
+        });
+        Ok(())
+    }
+
+    fn try_remove_register(username: UserName, registrant: ZkAddressType) -> DispatchResult {
+        // Username checks
+        username_validation(&username).ok_or(Error::<T>::InvalidUsernameFormat)?;
+
+        ensure!(
+            UsernameRecords::<T>::contains_key(&username),
+            Error::<T>::NotRegistered
+        );
+
+        ensure!(
+            UsernameRecords::<T>::get(&username).unwrap() == registrant,
```

Comment:
unwrap()?

Reply:
Ah checked above could be a `match` arm, but fine as is.

Maybe add comment above this one as well